domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java

```diff
@@ -1596,91 +1595,207 @@ public TotaledResults(ResultSet set, Integer total) {
+    boolean returnIdField, String distinctOn, List<FacetField> facets,
+    Handler<AsyncResult<PostgresClientStreamResult<T>>> replyHandler) {
 
-            sqlRowStream.resultSetClosedHandler(v -> sqlRowStream.moreResults()).handler(r -> {
-              JsonObject row = convertRowStreamArrayToObject(sqlRowStream, r);
-              try {
-                streamHandler.handle((T) deserializeRow(resultsHelper, externalColumnSetters, isAuditFlavored, row));
-              } catch (Exception e) {
-                sqlRowStream.close();
-                if (!transactionMode) {
-                  connection.close();
-                }
-                log.error(e.getMessage(), e);
-                replyHandler.handle(Future.failedFuture(e));
-              }
-            }).endHandler(v2 -> {
-              if (!transactionMode) {
-                connection.close();
-              }
-              replyHandler.handle(Future.succeededFuture());
-            });
+    try {
+      QueryHelper queryHelper = buildQueryHelper(table,
+        fieldName, wrapper, returnIdField, facets, distinctOn);
+
+      connection.querySingle(queryHelper.countQuery, countQueryResult -> {
+        if (countQueryResult.failed()) {
+          replyHandler.handle(Future.failedFuture(countQueryResult.cause()));
+          return;
+        }
+        ResultInfo resultInfo = new ResultInfo();
+        resultInfo.setTotalRecords(countQueryResult.result().getInteger(0));
+        doStreamGetQuery(connection, queryHelper.selectQuery, resultInfo,
+          clazz, facets, replyHandler);
+      });
+    } catch (Exception e) {
+      log.error(e.getMessage(), e);
+      replyHandler.handle(Future.failedFuture(e));
+    }
+  }
 
+  private <T> void doStreamGetQuery(SQLConnection connection, String selectQuery,
+    ResultInfo resultInfo, Class<T> clazz, List<FacetField> facets,
+    Handler<AsyncResult<PostgresClientStreamResult<T>>> replyHandler) {
+
+    connection.queryStream(selectQuery, stream -> {
+      if (stream.failed()) {
+        log.error(stream.cause().getMessage(), stream.cause());
+        replyHandler.handle(Future.failedFuture(stream.cause()));
+        return;
+      }
+      PostgresClientStreamResult<T> streamResult = new PostgresClientStreamResult(resultInfo);
+
+      SQLRowStream sqlRowStream = stream.result();
+      Promise<PostgresClientStreamResult<T>> promise = Promise.promise();
+      ResultsHelper<T> resultsHelper = new ResultsHelper<>(sqlRowStream, clazz);
+      boolean isAuditFlavored = isAuditFlavored(resultsHelper.clazz);
+      Map<String, Method> externalColumnSetters = getExternalColumnSetters(resultsHelper.columnNames,
+        resultsHelper.clazz, isAuditFlavored);
+      // fetch rows and produce ResultInfo when facets are all read
+      sqlRowStream.resultSetClosedHandler(v -> sqlRowStream.moreResults()).handler(r -> {
+        JsonObject row = convertRowStreamArrayToObject(sqlRowStream, r);
```

Comment:
Can you move this line (`JsonObject row = convertRowStreamArrayToObject(...`= into the `try` clause below so that any exception that `convertRowStreamArrayToObject` throws will get caught and handled?

Reply:
Sure